lib/ansible/plugins/lookup/onepassword.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (c) 2018, Scott Buchanan <sbuchanan@ri.pn>
+# Copyright (c) 2016, Andrew Zenk <azenk@umn.edu> (lastpass.py used as starting point)
+# Copyright (c) 2018, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = """
+    lookup: onepassword
+    author:
+      -  Scott Buchanan <sbuchanan@ri.pn>
+      -  Andrew Zenk <azenk@umn.edu>
+    version_added: "2.6"
+    requirements:
+      - op (1Password command line utility: https://support.1password.com/command-line/)
+      - must have already logged into 1Password using op CLI
+    short_description: fetch data from 1Password
+    description:
+      - use the op command line utility to fetch specific fields from 1Password
+    options:
+      _terms:
+        description: identifier(s) of object(s) from which you want to retrieve the field (UUID, name or domain)
+        required: True
+      field:
+        description: field to return from 1Password
+        default: 'password'
+      vault:
+        description: vault to retrieve from (if absent will search all available vaults)
+        default: None
+"""
+
+EXAMPLES = """
+- name: "retrieve password from entry named KITT in any vault"
+  debug: password="{{ lookup('onepassword', 'KITT') }}"
+
+- name: "retrieve username from entry named HAL 9000 in vault Discovery"
+  debug: password="{{ lookup('onepassword', 'HAL 9000', field='username', vault='Discovery') }}"
+"""
+
+RETURN = """
+  _raw:
+    description: field data requested
+"""
+
+import json
+from subprocess import Popen, PIPE
+
+from ansible.errors import AnsibleError
+from ansible.plugins.lookup import LookupBase
+
+
+class OnePassException(AnsibleError):
+
+    pass
+
+
+class OnePass(object):
+
+    def __init__(self, path='op'):
+        self._cli_path = path
+
+    @property
+    def cli_path(self):
+        return self._cli_path
+
+    def assert_logged_in(self):
+        try:
+            self._run(["get", "account"])
+        except:
+            raise OnePassException("Not logged into 1Password: please run 'op signin' first")
+
+    def get_field(self, item_id, field, vault=None):
+        args = ["get", "item", item_id]
+        if vault is not None:
+            args += ['--vault={0}'.format(vault)]
+        output, dummy = self._run(args)
+        return self._parse_field(field, output) if output != '' else ''
+
+    def _run(self, args, stdin=None, expected_rc=0):
+        p = Popen([self.cli_path] + args, stdout=PIPE, stderr=PIPE, stdin=PIPE)
+        out, err = p.communicate(stdin)
+        rc = p.wait()
+        if rc != expected_rc:
+            raise OnePassException(err)
+        return out, err
+
+    def _parse_field(self, field_name, data_json):
```

Comment:
I'm wondering if we should even parse the output here. From looking at a few items in my 1Password keychain, some of the `sections` don't have passwords in them. It may be better to simply return the entire contents of the item as JSON and allow the user to manipulate it in Ansible.


```
> op get item Dropbox | jq '.details.sections'
[
  {
    "name": "linked items",
    "title": "Related Items"
  },
  {
    "fields": [
      {
        "k": "concealed",
        "n": "TOTP_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
        "t": "",
        "v": "otpauth://totp/Dropbox?secret=XXXXXXXXXXXXXXXXXXXXXXXXXX"
      }
    ],
    "name": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "title": "Section"
  }
]
```

Reply:
I'd really push back against that suggestion. Ansible does not provide a way to easily parse complex and variable JSON structures. I could see a use case for a separate `raw` plugin or mode that would do what you describe, but I believe querying for a specific field (it doesn't have to be `password`) is sufficient and much cleaner in the playbook.

Reply:
I agree that parsing variable JSON structures is pretty tough inside a playbook — it's much better to do those type of things in Python. There are some filters, e.g., `map` and `extract`, that can do some data manipulation, but that's not what I was thinking.

My thought was that returning more data, maybe normalized into a regular format, rather than only a single value may be more useful in the long run. For instance, if the lookup returned `{username: 'user', password:'password', totp: '1234567', url: 'https://login.acme.com' }` it could be referenced as `{{ some_var.username }}` in the playbook. This may be too difficult to achieve given the widely varying nature of how things are stored in the 1Password database.

Reply:
I'm already ignoring sections as it is, just returning the first field that matches the provided field name. That's maybe not ideal, but it was an intentional tradeoff for simplicity. A fully flattened structure like you mention would mean doubling down on that, though, as I'd have to pick an approach like first or last field found for any given key. Alternately, leaving a hierarchical structure like `{{ some_var['Section Name'].username }}` is extra verbose.